### ISSUE TITLE
Add virtual methods for inheritance

### ISSUE THREAD

**Issue by MikePrime21** [restsharp/RestSharp#1783]
I have a really big rest api, so I've broken the client into multiple abstractions (ie... CustomerAbstraction, OrderAbstraction, ...).  I inherit from RestClient to create MyRestClient that gets injected into each abstraction.  All that works fine.  The problem is I want to add some base functionality in MyRestClient.  

Every response from this api is of type ApiResponse<RealObjectType>.  That is because the api serializes exceptions and passes them to the client.  I want to override the ExecuteAsync method so that it can receive the ApiResponse<> object and then throw any serialized exception from the server on the client.  I would return the RealObjectType if there was no error, thus hiding the ApiResponse entirely.  Unfortunately ExecuteAsync is not virtual so I can't do this.

**Comment by alexeyzimarev** [comment]
Why do you need to inherit if you can compose?

```charp
class MyRestClient {
    readonly RestClient _client;

    public MyRestClient(RestClient client) => _client = client;

    // do whatever you want with your own client
```

Have you thought of something like that?

**Comment by MikePrime21** [comment]
Yes, I'm already doing that, I'm just going a step further.  Our api is not just big, but its broken into sections for scalability.  Right now we have 3 sections each with different base urls.  I inherit RestClient into three other classes that each use the correct base address.  I then inject them in like you did above.  That all works fine.  

```
    internal class RestCustomerAbstraction : RestAbstractionBase, ICustomerAbstraction
    {
        public RestActionAbstraction(AdminRestClient _adminRestClient, MgmtRestClient _mgmtRestClient) 
        {
             _adminRestClient = adminRestClient
             _mgmtRestClient= mgmtRestClient
        }

        public async Task<Customer> GetCustomerAsync()
        {
            var request = new RestRequest($"api/v3/Customer");
            var response = await _mgmtRestClient.GetAsync<ManagedResponse<Customer>>(request);
            if(response.HasError)
            {
                ThrowResponseError(response);
            }
            return response.Result;
        }

```

Each call I make to the api, I check if there is a serialized error and if so I throw the deserialzied exception.  I can also create an extension method like GetManagedResponseAsync<>().  Again no issue there.  But if I want something global and enforced 100% of the time no matter what extension method is used, then I would have to override the ExecuteAsync method.  This may not be a perfect example, but what if I want to add global resiliency  (ie... retry the call 3 times on failure).  What if I want to capture a 401 error code and flush my token cache, then retry the call again forcing it to get a new token.  What if I want a simple audit trail so that every time a rest call is made, it logs the transaction.  What if I want to add in custom metrics around every rest call to log performance.  Don't get me wrong all of these things can and possibly should be done in other ways, but right now handling them at the RestClient layer is not even an option.  With no interface and no virtual methods, this RestClient is really rigid and I don't understand why.  Do I really need a valid reason for some virtual methods?  Sometimes I don't need the best solution, I just need something that works with little complexity.
